### PR TITLE
Postgres IDENTITY Column support

### PR DESCRIPTION
**Pull request by mitsos1os**
### Description of change

Fixes: #5365
This PR adds the ability to create **IDENTITY** columns _(for **Postgres 10+!**)_ using the `PrimaryGeneratedColumn` decorator, instead of the default `SERIAL` that was being created so far.

ex:
```typescript
@Entity()
export class User {
  @PrimaryGeneratedColumn("identity")
  id!: number;
}
```

The user must be cautious not to use the `identity` option in the decorator when connecting to a Postgres <10 database because it will throw an error since it is not supported in the DB Engine.

I chose to add `identity` as an extra `generationStrategy` instead of adding an extra option flag, because it directly avoids rest of boilerplate code provided from the `increment` default strategy _(Adding Sequences)_. If extra flag was chosen and `increment` was kept as a `generationStrategy`, extra logic would have to be applied to check the flag's value and not create the **SERIAL** column with its associated **SEQUENCES** and **DEFAULT VALUES** accordingly.

Unit tests have been added that test the produced SQL for creating the column in a table.

However no functional tests added that can load the table from the database and apply proper `TableColumn` options _(although the realted code has been added)_, due to the fact that older Postgres version is used in docker-compose.yml. I could directly upgrade it in the compose file and apply my test, but I am not sure this is a wanted behavior.

Open to suggestions and comments!

Thanks!

### Pull-Request Checklist

- [x] Code is up-to-date with the `master` branch
- [x] `npm run lint` passes with this change
- [x] `npm run test` passes with this change
- [x] This pull request links relevant issues as `Fixes #0000`
- [x] There are new or updated unit tests validating the change
- [x] Documentation has been updated to reflect this change
- [x] The new commits follow conventions explained in [CONTRIBUTING.md](https://github.com/typeorm/typeorm/blob/master/CONTRIBUTING.md)

### CODE REVIEW

**Comment by imnotjames** [comment]
Why not use update things so that `increment` uses the `IDENTITY` syntax when connected to postgres 10?

**Comment by mitsos1os** [comment]
> Why not use update things so that `increment` uses the `IDENTITY` syntax when connected to postgres 10?

Hi! Can you please explain what you mean by saying **update things **? I didn't understand your suggestion

**Comment by imnotjames** [comment]
> > Why not use update things so that `increment` uses the `IDENTITY` syntax when connected to postgres 10?
> 
> Hi! Can you please explain what you mean by saying **update things**? I didn't understand your suggestion

Change the default behavior of `increment` for primary generated column so that it uses the new postgres feature when connected to postgres 10.

Is that possible?

**Comment by mitsos1os** [comment]
@imnotjames From what I have seen there is no information for the version of the Database Server to which the client is connected to. The only thing I found with a quick look is something like that https://github.com/brianc/node-postgres/issues/2002 which would require extra tampering with the embedded driver functionality.

Also I stated in my PR description that I chose a new generation strategy to also circumvent the accompanying SEQUENCE creation that comes with SERIAL column definition.

**Comment by imnotjames** [comment]
> @imnotjames From what I have seen there is no information for the version of the Database Server to which the client is connected to. The only thing I found with a quick look is something like that https://github.com/brianc/node-postgres/issues/2002 which would require extra tampering with the embedded driver functionality.

See https://github.com/typeorm/typeorm/pull/7716

> 
> Also I stated in my PR description that I chose a new generation strategy to also circumvent the accompanying SEQUENCE creation that comes with SERIAL column definition.

It seems that this is the new suggested mechanism for increment instead of serial when supported, right?

I don't think we should add yet another generation strategy that's custom for a specific database.  I also think we should drop `rowid` because cockroachdb seems to use that as the default anyway.

**Comment by mitsos1os** [comment]
Ok thanks! I will take a look.
Just to make sure, will you review this for integration?
What I mean is it would be good to avoid a different opinion from a reviewer that would want to revert this change again

If it is OK, I have no problem changing it to what you suggested

**Comment by mitsos1os** [comment]
OK, I took a look at the version checking, it is straightforward.
How should I proceed with my PR without re-writing the same code for version checking as [here](https://github.com/typeorm/typeorm/pull/7716/files#diff-b924b9a6d4ab4328f4ad59d4d96b794f6a1975fae8d1d3708909e31c151730b9R1928)?

**Comment by imnotjames** [comment]
> Ok thanks! I will take a look.
> Just to make sure, will you review this for integration?
> What I mean is it would be good to avoid a different opinion from a reviewer that would want to revert this change again
> 
> If it is OK, I have no problem changing it to what you suggested

I can review it and I have the ability to merge code.

I can't 100% say that it will be merged in.

We can always leave this PR open, create a new branch, get that code working, and then make sure the tests show things are good.

The only thing I'm unsure about is what will happen with existing tables when this change gets in.  I think the answer is "nothing" - they'll stay the same, but if they are recreated they would get the new code / sequence / etc.  That's probably fine?

**Comment by mitsos1os** [comment]
@imnotjames OK cool!
Regarding the #7716 PR where the `getVersion` is implemented, how do you want to handle this since I will have to use the same/similar util?
Should I also create it in my PR?

Also left a comment regarding my view on the implementation [here](https://github.com/typeorm/typeorm/pull/7716#discussion_r654630750)

**Comment by imnotjames** [comment]
If you need it, write it.  If it gets merged in first we'll update that PR to match.

File: docs/entities.md [hunk 1]
```diff
@@ -524,7 +524,7 @@ export class User {
 
 `uuid` value will be automatically generated and stored into the database.
 
-Besides "uuid" there is also "increment" and "rowid" (CockroachDB only) generated types, however there are some limitations
+Besides "uuid" there is also "increment" "identity" (Postgres 10+ only) and "rowid" (CockroachDB only) generated types, however there are some limitations
```
Comment (pleerock) [inline docs/entities.md:527]: missing "," mark between `"increment" "identity"`

File: src/driver/postgres/PostgresQueryRunner.ts [hunk 2]
```diff
@@ -930,7 +930,7 @@ export class PostgresQueryRunner extends BaseQueryRunner implements QueryRunner
                 }
             }
 
-            if (oldColumn.isGenerated !== newColumn.isGenerated && newColumn.generationStrategy !== "uuid") {
+            if (oldColumn.isGenerated !== newColumn.isGenerated && !["uuid", "identity"].includes(newColumn.generationStrategy as string)) {
```
Comment (pleerock) [inline src/driver/postgres/PostgresQueryRunner.ts:933]: Simply this expression. My stupid brain can't accept too tricky expressions 😆

**Review by pleerock — CHANGES_REQUESTED** [review]

**Comment by mitsos1os** [comment]
@pleerock thanks for getting back! I have rebased and addressed the comments on your conversations. Take a look and let me know what you think!
Thanks!

**Comment by mitsos1os** [comment]
Great! Looking forward to it getting merged!

**Pull request merged**